### ISSUE TITLE
Preserve "\r" even when not using a terminal (ex:redirecting to a file)

### ISSUE THREAD

**Issue by sebma** [mpv-player/mpv#6617]
### mpv version and platform
2:0.29.1+git2~trusty on Ubuntu trusty

### Reproduction steps
```
$ mpv --no-resume-playback https://www.facebook.com/TBN/videos/1580372468665943/ >mpv.log 2>&1
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 50s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 50s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 54s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 59s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 61s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 63s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 66s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 69s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 72s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 75s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 78s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 83s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 85s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 89s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 91s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 94s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 96s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 102s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 106s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 109s+10MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 111s+10MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 115s+10MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 119s+11MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 123s+11MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 128s+11MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 134s+12MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 138s+12MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 142s+12MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 146s+13MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 150s+13MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 154s+13MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 156s+13MB
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 159s+14MB
Saving state.


Exiting... (Quit)
```

### Expected behavior
```
$ mpv --no-resume-playback https://www.facebook.com/TBN/videos/1580372468665943/ >mpv.log 2>&1
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p
AV: 00:00:01 / 00:56:50 (0%) A-V:  0.000 Cache: 159s+14MB
Saving state.


Exiting... (Quit)
```

### Actual behavior
The "\r" that are output at the end "AV:" lines when writing stderr to a terminal are replaced by "\n" when writing <s>stderr</s> to a <s>pipe or a</s> file.

Can you change this behavior by default or can you please add an option for that ?

**Comment by zc62** [comment]
Then what's the point of displaying that line. Just use [`--quiet`](https://mpv.io/manual/master/#options-quiet) to mute it.

**Comment by sebma** [comment]
I want to be able to keep the last position given by the last `AV:` line in my `mpv` logs.

Why did you bother to write to terminal or files differently anyway ?

Or has this something to do with the operating system treating them differently ?

**Comment by qyot27** [comment]
Pipes are like time, they only flow in one direction; once data is written to the pipe, it cannot be backtracked inside the pipe and updated in-place.  It's not mpv doing any of this, nor is it the operating system.  It's how pipes work.  Your only option is filtering the extraneous lines out of the piped files yourself by grep/sed/awk/what-have-you:
```
#!/bin/bash
LOGFILE_MAIN="$(echo $1_$(date -d now +"%s").log)"

mpv $1 2>&1 | tee $LOGFILE_MAIN
INPUTTIME="$(grep "A\:\|V\:\|AV\:" "$LOGFILE_MAIN" | tail -1)"
LASTLINE="$(tail -1 "$LOGFILE_MAIN")"
grep -v "A\:\|V\:\|AV\:" $LOGFILE_MAIN | head -n -1 > "$LOGFILE_MAIN"_notime.log
echo $INPUTTIME >> "$LOGFILE_MAIN"_notime.log
echo $LASTLINE >> "$LOGFILE_MAIN"_notime.log
mv "$LOGFILE_MAIN"_notime.log $LOGFILE_MAIN
```

**Comment by sebma** [comment]
@qyot27 Ok, my mistake, I shouldn't have written about pipes because pipes can't be rewinded so let's forget about pipes one moment.

I my first post I mentioned that `mpv` was using `\n` at the end of each `AV:` line when writing to a log :
```
$ mpv --no-resume-playback https://www.facebook.com/TBN/videos/1580372468665943/ >toto.log 2>&1
$ # AND IN ANOTHER TERMINAL WINDOW : 
$ cat toto.log
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p
AV: 00:00:00 / 00:56:50 (0%) A-V: -0.000 Cache: 44s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 45s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 48s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 52s+4MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 55s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 57s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 60s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 62s+5MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 65s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 69s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 71s+6MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 75s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 77s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 80s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 83s+7MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 89s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 91s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 96s+8MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 99s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 103s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 107s+9MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 110s+10MB
AV: 00:00:00 / 00:56:50 (0%) A-V:  0.000 Cache: 113s+10MB
Saving state.


Exiting... (Quit)
```
 as opposed to using `\r` at the end of each `AV:` line when writing to the terminal : 
```
$ mpv --no-resume-playback https://www.facebook.com/TBN/videos/1580372468665943/
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p
AV: 00:00:02 / 00:56:50 (0%) A-V:  0.000 Cache: 392s+33MB
Saving state.


Exiting... (Quit)
```

Why not use `\r` at the end of each `AV:` line in both cases, is it not technically possible ?

**Comment by qyot27** [comment]
The shell uses the same stream-of-characters-in-text-mode concept to write files from redirected output as it does to pipe data to stdin.  And similarly, just because the shell can interpret a carriage return correctly inside the Terminal buffer, doesn't mean it won't treat it as a line terminator when redirected.

https://unix.stackexchange.com/questions/374592/carriage-return-with-echo-command

The example script in the answer exhibits both of the same problems; when executed normally, it ticks the time upward on the same line.  When redirected, the changed times all get written to new lines.  And even moreso, here on MSys2's bash environment, the `cat -A` output on the written logfile is this:
```
^MSun, Apr 7, 2019 10:25:30 PM^MSun, Apr 7, 2019 10:25:31 PM^MSun, Apr 7, 2019 10:25:32 PM
```
Just carriage returns, no line feeds.  And opening it in Notepad or Notepad2 show that it's plain CR line endings (with Notepad even making reference to the fact that singular CR as a line ending is for Macintosh).

**Comment by sebma** [comment]
@qyot27 Maybe I'm being a little ambiguous, so I wrote a script which I hope can help myself be understood : 
```bash
$ cat mpv_AV_test.bash
#!/usr/bin/env bash

cat <<-EOF
Playing: $1
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p"
EOF

trap 'echo -e "\nSaving state.\n\n\nExiting... (Quit)";exit' INT

i=0
seconds=0
while [ $seconds != 60 ]
do
	printf 'AV: 00:00:%02d / 00:56:50 (0%%) A-V:  0.000 Cache: 1767s+147MB\r' $i
	let i++
	sleep 1
done

trap - INT
```
Here is the ouput when run from a terminal, as you can see I get all the output but with only the last `AV:` position : 
```bash
$ mpv_AV_test.bash https://www.facebook.com/TBN/videos/1580372468665943/
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p"
^C: 00:00:04 / 00:56:50 (0%) A-V:  0.000 Cache: 1767s+147MB
Saving state.


Exiting... (Quit)
```
And if I redirect the output of the script to a file, the output is the same, `\r` are preserved, there are not replaced by `\n` : 
```
$ mpv_AV_test.bash https://www.facebook.com/TBN/videos/1580372468665943/ > toto.log 2>&1
^C
$ cat toto.log 
Playing: https://www.facebook.com/TBN/videos/1580372468665943/
 (+) Video --vid=1 (*) (h264 634x360 30.000fps)
 (+) Audio --aid=1 (*) (aac 1ch 48000Hz)
AO: [pulse] 48000Hz mono 1ch float
VO: [gpu] 634x360 yuv420p"
AV: 00:00:05 / 00:56:50 (0%) A-V:  0.000 Cache: 1767s+147MB
Saving state.


Exiting... (Quit)
```
Does this help to understand what my feature request is ?

**Comment by DanOscarsson** [comment]
I missed that somebody wanted this.
They way it worked by using terminal semantics when writing to terminal and file semantics when to a file was very good. By having \n instead or \r in the file, the log file could easily be searched in with grep or view in less or an editor.
If I remember \r (and terminal codes) were written to files too and I have a special patch to fix it. When it was changed to write \n when sending to files, my patch was no longer needed.
Now I will need one again.

**Comment by wm4** [comment]
What a load of bullshit. You can still use --quiet or --log-file. Jesus Christ.

**Comment by DanOscarsson** [comment]
Neither of those help. I want to have all that is output on terminal to a file. Like status lines and same log level as is used on terminal. Otherwise it is much more difficult to trace problems.
And there are other programs that do things differently with a terminal then when redirected to a file. For example: less.
If we can have --log-file that includes both status lines and logs depending on a defined log-level so I can get the same output that I can see on the terminal in the log file (including all sent to stderr output), then the log file would do.

**Comment by sebma** [comment]
wm4 fixed it in v0.30.0 : 
```bash
$ git describe --contains 8721ac5
v0.30.0~10
$ git log v0.29.1..v0.30.0 | sed -n '/ 8721ac5/,/^commit/p'
commit 8721ac50fb7d779476d7cad0632cdb69fdc100e3
Author: wm4 <wm4@nowhere>
Date:   Thu Oct 24 13:13:12 2019 +0200

    msg: always use terminal control codes for status line
    
    Before this commit, the status line used terminal control codes only if
    stderr was a terminal. I'm not sure why this was done, and git blame
    tracks it back to a huge commit by me, which changed all of the terminal
    handling.
    
    A user complained, so just stop treating this specially for no reason.
    
    Fixes: #6617

commit a935109235b1032fe15a6ce790fafbcd401ecf6e
```
